media/starboard/starboard_cdm.cc

```diff
@@ -12,371 +12,323 @@
-                  << wrapped_drm_system_ << ") with type: " << type
-                  << ", message size: " << message_size;
-
-        session_update_request.generated_callback.Run(type, std::move(message),
-                                                      message_size);
-      } else {
-        // Failure during request generation.
-        LOG(INFO) << "Calling session update request callback on drm system ("
-                  << wrapped_drm_system_ << "), status: " << status
-                  << ", error message: " << error_message;
-        session_update_request.did_not_generate_callback.Run(status,
-                                                             error_message);
-      }
-    }
 
-    // Sweep the context of |GenerateSessionUpdateRequest| once license updated.
+    auto session_request = std::move(session_update_request_iterator->second);
+    session_request.session->SetSessionId(session_id.value());
+    session_map_[session_id.value()] = std::move(session_request.session);
+
     ticket_to_session_update_request_map_.erase(
         session_update_request_iterator);
-  } else {
-    // Called back spontaneously by the underlying DRM system.
 
-    // Spontaneous calls must refer to a valid session.
-    if (!session_id) {
-      LOG(ERROR) << "SbDrmSessionUpdateRequestFunc() should not be called "
-                    "with both invalid ticket and null session id.";
-      NOTREACHED();
-      return;
-    }
+    promises_.ResolvePromise(session_request.promise_id, session_id.value());
+  }
 
-    // Find the session by ID.
-    IdToSessionMap::iterator session_iterator =
-        id_to_session_map_.find(*session_id);
-    if (session_iterator == id_to_session_map_.end()) {
-      LOG(ERROR) << "Unknown session id: " << *session_id << ".";
-      return;
-    }
+  LOG(INFO) << "Calling session update request callback on drm system ("
+            << sb_drm_ << ") with type: " << type
+            << ", message size: " << message.size();
 
-    // As DrmSystem::Session may be released, need to check it before using it.
-    if (session_iterator->second) {
-      LOG(INFO) << "Calling session update request callback on drm system ("
-                << wrapped_drm_system_ << "), type: " << type
-                << ", message size " << message_size;
-      session_iterator->second->update_request_generated_callback().Run(
-          type, std::move(message), message_size);
-    }
+  if (!session_id.has_value()) {
+    LOG(ERROR) << "No session id provided";
+    return;
   }
+  auto session_iterator = session_map_.find(session_id.value());
+  if (session_iterator == session_map_.end()) {
+    LOG(ERROR) << "Unknown session id: " << session_id.value() << ".";
+    return;
+  }
+  StarboardCdmSession* session = session_iterator->second.get();
+  session->InvokeOnSessionMessage(
+      SbDrmSessionRequestTypeToMediaMessageType(type), message);
```

Comment:
This call will call into `StarboardCdmSession::InvokeOnSessionMessage()`, which will call `StarboardCdmSession::OnSessionMessage()`, and ended up calling the member variable `message_cb_` of this class with `session_id_`.  The `session_id_` should be the same as the `session_id` above.

Just wanted to double check if there are any considerations for this extra forwarding, than calling `message_cb_` directly here (maybe need to be bound with current loop)?

Reply:
ah, good call. I managed to simplify the implementation by binding here.